// @ethersphere/bee-js/dist/mjs/modules/debug/states.js
var __awaiter = this && this.__awaiter || function (thisArg, _arguments, P, generator) {
  function adopt(value) {
    return value instanceof P ? value : new P(function (resolve) {
      resolve(value);
    });
  }

  return new (P || (P = Promise))(function (resolve, reject) {
    function fulfilled(value) {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    }

    function rejected(value) {
      try {
        step(generator["throw"](value));
      } catch (e) {
        reject(e);
      }
    }

    function step(result) {
      result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
    }

    step((generator = generator.apply(thisArg, _arguments || [])).next());
  });
};

import { http } from "../../utils/http.js";
const RESERVE_STATE_ENDPOINT = 'reservestate';
const WALLET_ENDPOINT = 'wallet';
const CHAIN_STATE_ENDPOINT = 'chainstate';
/**
 * Get state of reserve
 *
 * @param ky Ky debug instance
 */

export function getReserveState(ky) {
  return __awaiter(this, void 0, void 0, function* () {
    const response = yield http(ky, {
      method: 'get',
      path: `${RESERVE_STATE_ENDPOINT}`,
      responseType: 'json'
    });
    return response.data;
  });
}
/**
 * Get state of reserve
 *
 * @param ky Ky debug instance
 */

export function getChainState(ky) {
  return __awaiter(this, void 0, void 0, function* () {
    const response = yield http(ky, {
      method: 'get',
      path: `${CHAIN_STATE_ENDPOINT}`,
      responseType: 'json'
    });
    return response.data;
  });
}
/**
 * Get wallet balances for xDai and BZZ of the node
 *
 * @param ky Ky debug instance
 */

export function getWalletBalance(ky) {
  return __awaiter(this, void 0, void 0, function* () {
    const response = yield http(ky, {
      method: 'get',
      path: `${WALLET_ENDPOINT}`,
      responseType: 'json'
    });
    return response.data;
  });
}
