// @ethersphere/bee-js/dist/cjs/modules/debug/states.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWalletBalance = exports.getChainState = exports.getReserveState = void 0;
const http_1 = require("../../utils/http");
const RESERVE_STATE_ENDPOINT = 'reservestate';
const WALLET_ENDPOINT = 'wallet';
const CHAIN_STATE_ENDPOINT = 'chainstate';
/**
 * Get state of reserve
 *
 * @param ky Ky debug instance
 */
function getReserveState(ky) {
    return __awaiter(this, void 0, void 0, function* () {
        const response = yield (0, http_1.http)(ky, {
            method: 'get',
            path: `${RESERVE_STATE_ENDPOINT}`,
            responseType: 'json',
        });
        return response.data;
    });
}
exports.getReserveState = getReserveState;
/**
 * Get state of reserve
 *
 * @param ky Ky debug instance
 */
function getChainState(ky) {
    return __awaiter(this, void 0, void 0, function* () {
        const response = yield (0, http_1.http)(ky, {
            method: 'get',
            path: `${CHAIN_STATE_ENDPOINT}`,
            responseType: 'json',
        });
        return response.data;
    });
}
exports.getChainState = getChainState;
/**
 * Get wallet balances for xDai and BZZ of the node
 *
 * @param ky Ky debug instance
 */
function getWalletBalance(ky) {
    return __awaiter(this, void 0, void 0, function* () {
        const response = yield (0, http_1.http)(ky, {
            method: 'get',
            path: `${WALLET_ENDPOINT}`,
            responseType: 'json',
        });
        return response.data;
    });
}
exports.getWalletBalance = getWalletBalance;
